### PR TITLE
shim: Handle a SIGKILL signal

### PR DESCRIPTION
**Pull request by sboeuf**
Because a SIGKILL cannot be caught from the targeted process, we
had to find a way to detect the end of this process.
This patch proposes a solution relying on a child that we spawn
on behalf of the parent process. The only purpose of that child
is to listen to a monitoring pipe that has been opened from the
parent. In case the child receives no bytes after he was waiting
on this pipe, this means the parent could not write to the pipe
before to exit, meaning we have detected a case of SIGKILL. Then,
the child forward this signal to the container process through the
proxy, and it exits immediately with a call to _exit() system call.
This allows the child to avoid calling into the parent atexit()
functions.

### CODE REVIEW

**Comment by sboeuf** [comment]
Fixing https://github.com/clearcontainers/shim/issues/38

File: src/shim.c [hunk 1]
```diff
@@ -1090,6 +1105,55 @@ main(int argc, char **argv)
...
+			shim_debug("Parent has terminated because of SIGKILL"
+				"/nForwarding the SIGKILL to the container"
+				" process");
+
+			signal_handler(SIGKILL);
+			handle_signals(&shim);
+
+			/* restore the terminal because atexit functions won't
+			 * be called
+			 */
```
Comment (amshinde) [inline src/shim.c]: Why won't the atexit function called as you are exiting with exit()

Comment (sboeuf) [reply src/shim.c]: I am exiting with `_exit()` which exits immediately: http://man7.org/linux/man-pages/man2/_exit.2.html

Comment (amshinde) [reply src/shim.c]: ok. thanks for clarifying. Would be useful to add the "with _exit()" to your above comment.
Other than that lgtm

**Comment by amshinde** [comment]
lgtm

[Approved with PullApprove](https://pullapprove.com/clearcontainers/shim/pull-request/39/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=clearcontainers/shim) [Approved with PullApprove](https://pullapprove.com/clearcontainers/shim/pull-request/39/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=clearcontainers/shim)

**Comment by jcvenegas** [comment]
lgtm , @sboeuf  @amshinde  I think this should be ported to 2.1 as well 

[Approved with PullApprove](https://pullapprove.com/clearcontainers/shim/pull-request/39/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=clearcontainers/shim)

File: src/shim.c (hunk 1 above)
Comment (sboeuf) [reply src/shim.c]: Yes you're right, let me add this to the comment

**Comment by sboeuf** [comment]
@jcvenegas according to the minimal amount of work implied by this change, I agree this should be ported to 2.1.

**Pull request merged**